File: frontend/src/modules/editor/contentProjection.ts
/**
 * Normalize editor-only JSON differences before comparing a local Yjs
 * document with the relational projection. Tiptap may add block ids and
 * default null attributes while mounting; those are useful in the editor but
 * are not a user edit and must not make a navigation/action barrier wait
 * forever for a byte-for-byte match that can never occur.
 */
function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function canonicalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonicalize)
  if (!isRecord(value)) return value

  const result: Record<string, unknown> = {}
  for (const key of Object.keys(value).sort()) {
    const child = value[key]
    if (key === 'attrs' && isRecord(child)) {
      const attrs: Record<string, unknown> = {}
      for (const attrKey of Object.keys(child).sort()) {
        if (attrKey === 'blockId' || child[attrKey] === null) continue
        attrs[attrKey] = canonicalize(child[attrKey])
      }
      if (Object.keys(attrs).length > 0) result[key] = attrs
      continue
    }
    if (key === 'content' && Array.isArray(child)) {
      if (child.length > 0) result[key] = child.map(canonicalize)
      continue
    }
    result[key] = canonicalize(child)
  }
  return result
}

export function contentJsonSemanticallyEqual(left: unknown, right: unknown): boolean {
  return JSON.stringify(canonicalize(left)) === JSON.stringify(canonicalize(right))
}

File: backend/app/services/content_projection.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any


def _canonical(value: Any) -> Any:
    """Normalize editor-only JSON noise for projection no-op checks.

    Tiptap can add default ``null`` attributes and stable block ids while an
    editor is mounting. Those changes are useful inside Yjs, but they are not
    a user edit and must not trigger a task/note notification merely because a
    document was opened. Identity attributes such as taskId/noteId remain
    significant; only blockId is intentionally ignored here.
    """
    if isinstance(value, Mapping):
        result: dict[str, Any] = {}
        for key, child in value.items():
            if key == "attrs" and isinstance(child, Mapping):
                attrs = {
                    attr_key: _canonical(attr_value)
                    for attr_key, attr_value in child.items()
                    if attr_key != "blockId" and attr_value is not None
                }
                if attrs:
                    result[key] = attrs
                continue
            if key == "content" and isinstance(child, list):
                normalized_content = [_canonical(item) for item in child]
                if normalized_content:
                    result[key] = normalized_content
                continue
            result[key] = _canonical(child)
        return result
    if isinstance(value, list):
        return [_canonical(item) for item in value]
    return value


def content_json_semantically_equal(left: Mapping[str, Any] | None, right: Mapping[str, Any] | None) -> bool:
    return _canonical(left) == _canonical(right)
